lib/source_file.cpp

```diff
@@ -226,6 +226,36 @@ cstring InputSources::getSourceFragment(const SourceInfo &position) const {
     return result + toadd + marker + cstring::newline;
 }
 
+cstring InputSources::getSourceFragment2(const SourceInfo &position) const {
```

Comment:
I haven't had a chance to really review this, but one initial comment is that this method needs a more descriptive name, and ideally a Doxygen comment describing what it does.